light-client/verifier/src/verifier.rs

```diff
@@ -0,0 +1,117 @@
+    pub historical_validator_sets: Vec<(Range<BlockIndex>, TrustedValidatorSet)>,
+    /// A list of known valid block ids, which may appear before
+    /// trusted_validator_set_start_block.
+    pub known_valid_block_ids: HashSet<BlockID>,
+}
+
+// For a simple LightClientVerifier initialization
+impl From<TrustedValidatorSet> for LightClientVerifier {
+    fn from(src: TrustedValidatorSet) -> Self {
+        Self {
+            trusted_validator_set: src,
+            trusted_validator_set_start_block: 0,
+            historical_validator_sets: Default::default(),
+            known_valid_block_ids: Default::default(),
+        }
+    }
+}
+
+impl LightClientVerifier {
+    /// Validate that a given block has been externalized by the network, given
+    /// additional evidence in the form of BlockMetadata with signatures.
+    pub fn validate_block(&self, block: &Block, metadata: &[BlockMetadata]) -> Result<(), Error> {
+        if !block.is_block_id_valid() {
+            return Err(Error::InvalidBlockId);
+        }
+        if self.known_valid_block_ids.contains(&block.id) {
+            return Ok(());
+        }
+        if block.index >= self.trusted_validator_set_start_block {
+            let metadata_block_id = self
+                .trusted_validator_set
+                .validate_block_id_signatures(metadata)?;
+            if metadata_block_id != block.id {
+                return Err(Error::BlockIdMismatch);
+            }
+            return Ok(());
+        }
+
+        for (range, validator_set) in self.historical_validator_sets.iter() {
+            if range.contains(&block.index) {
+                let metadata_block_id = validator_set.validate_block_id_signatures(metadata)?;
+                if metadata_block_id != block.id {
+                    return Err(Error::BlockIdMismatch);
+                }
+                return Ok(());
+            }
+        }
+        Err(Error::NoMatchingValidatorSet)
+    }
+
+    /// Validate that a given block has been externalized, and that it matches
+    /// to given block contents
+    pub fn validate_block_and_block_contents(
+        &self,
+        block: &Block,
+        block_contents: &BlockContents,
+        block_metadata: &[BlockMetadata],
+    ) -> Result<(), Error> {
+        // Validate the block
+        self.validate_block(block, block_metadata)?;
+        // Validate that the contents match the block
+        let contents_hash = block_contents.hash();
+        if contents_hash != block.contents_hash {
+            return Err(Error::BlockContentHashMismatch);
+        }
+        Ok(())
+    }
+
+    /// Validate that one or more TxOut's appeared in a particular block that
+    /// was externalized
+    pub fn validate_txos_in_block(
+        &self,
+        txos: &[TxOut],
+        block: &Block,
+        block_contents: &BlockContents,
+        block_metadata: &[BlockMetadata],
+    ) -> Result<(), Error> {
+        // Validate the block and block contents
```

Comment:
:upside_down_face: Suggest omitting this comment it seems to be reiterating the function name below, communicating "what" and not providing any extra insight for "why"
```suggestion
```